DataDog/datadog-agent
# Split IoT Agent main and refactor constants

### What does this PR do?

- Create `flavor` package and move constants for agent names/flavors to this module
- Move Windows Agent service to new package
- Create IoT Agent main in Linux and Windows
- Set `flavor.AgentFlavor` in the main of every agent
- Remove `agentName` from `BufferedAggregator`, `InitAggregator` and `InitAggregatorWithFlushInterval`.

### Motivation

- Remove duplication and simplify code.
- Move towards a model where different Agents can reuse common code.
- Have a more robust way to distinguish between IoT and full Agent. Right now, if someone does a custom build and forgets the linker flag we might get an incorrect flavor. This way there is only one thing to change: the main package.

### Additional Notes

- The vast majority of changes are modifying calls to `InitAggregator`. Since the `BufferedAggregator` doesn't need the `agentName` field anymore, all calls to `InitAggregator` and related functions can be simplified.
- Reverts #5754 and related PRs since we don't use the linker flag anymore.
- `flavor.AgentFlavor` must not be used before the main function gets called, since that sets the value. This includes package init functions and package variables and constants initialization.

### Describe your test plan

There should be no visible user-facing changes so we should just:
- check the agent runs correctly on:
     * [x] Windows as a service
     * [x] Linux IoT
     * [x] Linux Full Agent
     * [X] ~~Android~~ (no longer changed in this PR) 

- check the `status` output is correct for:
     * [X] Cluster Agent
     * [X] Full Agent
     * [x] IoT Agent
     * [x] DogstatsD

- check the `iot_host` config value is set to true only for the IoT Agent
